### DISCUSSION TITLE
Alternative approach to large playlist of local files or play files from a local folder

### DISCUSSION THREAD

**Discussion by donburch888** [music-assistant/support#2524] [Feature requests and ideas]
I have a fairly large number of tracks in my local collection, and wish to (random) play tracks from a selected folder (and sub-folders) ... kind of like a local radio station without adverts. 
I could build a playlist - but I expect that a list of 17,000 tracks would be unwieldy for Music Assistant, and for the client device which is asked to play the list. A different approach may be required as an alternative to large playlists. 

I seem to recall a suggestion for the server app to send the next file to the client whenever the currently playing track is about to end.  I note that M3U playlists generally include each track's playing time; and I expect that an estimate of playing time could be derived from the individual file properties ... or (of course) directly from the Music Assistant database.  
To look at it another way, whereas a playlist is a static list of files, this would be a dynamic service running in the background, periodically selecting and adding the next track to the client device's queue

I have noticed several other people mentioning large playlists or suggesting much the same thing for Home Assistant, so I believe I am not alone. 
[[Music Assistant 2.0: Your Music, Your Players](https://community.home-assistant.io/t/music-assistant-2-0-your-music-your-players/727626/472)
[Media Player to play a folder?](https://community.home-assistant.io/t/media-player-to-play-a-folder/424310)
[Local M3U Playlist Song Limit #2323](https://github.com/orgs/music-assistant/discussions/2323)
[New Folders category to mirror the filesystem tree #413](https://github.com/orgs/music-assistant/discussions/413)

**Comment by kepstin** [comment]
When trying out Music Assistant on my collection, the first thing I noticed is that I could not simply go to the All Tracks view and hit play. With most "Library" style music players, I'm able to shuffle my entire music collection (or, actually, play randomly - i.e. after each track select a (possibly weighted) random track to play next). It seems like there's no way to do that in Music Assistant other than to add all of my music to a single playlist - and adding 25k+ tracks in one playlist doesn't seem like a viable option?

Library players that support playing the library directly often also have a "Queue" feature - the way this usually works is that when the current track ends, then if the queue contains any tracks, it will take the first track from the queue and play that - and remove it from the queue once played. Otherwise if the queue is empty, the player picks the next track from the library in the configured order.

This is made a bit tricker from how the providers seem to work - due to details of my music collection organization, I have things spread over multiple folder and need multiple "Local Files" providers that together make up my local collection. I'd like to play tracks from all of those merged together - while excluding any cloud providers, radio providers, podcast providers, etc.

**Reply by marcelveldt** [reply]
We removed the feature to play ALL library tracks because we had folks with 500.000 tracks hitting that button to then run out of memory. Instead we have some nice randomized playlists that takes a selection of tracks (max 2500). Random and also sorted on less played.

**Reply by kepstin** [reply]
Yeah - that's why I'm used to players which don't require loading the entire library into a single playlist, but instead pick a single next track to play from the library on demand when the queue is empty.

**Reply by OzGav** [reply]
That sounds like the smart playlist feature that is being worked on. No ETA for that though. https://github.com/orgs/music-assistant/projects/2/views/1?pane=issue&itemId=70609106

**Reply by marcelveldt** [reply]
> Yeah - that's why I'm used to players which don't require loading the entire library into a single playlist, but instead pick a single next track to play from the library on demand when the queue is empty.

That is a whole different request that "play all tracks" but more like a dynamic radio - something we also have.
We just do not allow sending ALL TRACKS to the queue - we advise people to then just use the radio mode or one of the dynamic playlists.

So this feature you are talking about already exists, just hit the "start radio" button in the UI, next to an artist, album or playlist:

![Scherm­afbeelding 2024-09-05 om 00 33 36](https://github.com/user-attachments/assets/b40b8cc5-9d11-4b58-918c-2f068d4686ad)

**Reply by marcelveldt** [reply]
Only thing missing is good genre support but that will also come at some point and then we will also have the radio mode support

**Reply by kepstin** [reply]
Yeah - the missing piece to me (that made Radio mode undiscoverable) was that there's no way to start Radio mode unless you first select a specific artist, album, track, playlist, etc. I want to just "Start Radio" and have it pick the first track on its own.

**Reply by marcelveldt** [reply]
Sounds like a nice feature request to add some day, a "start radio" button within the tracks listing that starts radio mode from random tracks in your library.

**Reply by OzGav** [reply]
@kepstin in Integration 2024.11.0 there is a new action `get_library`. You could use this to do what you requested by using the following and then passing the random track to the `play_media` action with `radio_mode` enabled.

```
action: mass.get_library
data:
  media_type: track
  limit: 1
  order_by: random
response_variable: random_track
action: mass.play_media
data:
  media_id: "{{ random_track.tracks[0].uri }}"
  media_type: track
  enqueue: play
  radio_mode: true
target:
  entity_id: media_player.ma_kitchen_speaker
```

**Comment by OzGav** [comment]
Playing a folder from the browse view is available in 2.4.0b28

Between that, the get_library action, and the radio feature is there anything else in the feature request that isn't satisified?
